File: articles/applied-ai-services/form-recognizer/prebuilt-models/sample-code/python/prebuilt-invoice.2022-06-30-preview.py
import os
import sys
from azure.core.credentials import AzureKeyCredential
from azure.ai.formrecognizer import DocumentAnalysisClient

def analyze_document(args):
    endpoint = args[1] if len(args) > 1 else "<your-endpoint>"
    key = args[2] if len(args) > 2 else "<your-key>"
    doc_url = args[3] if len(args) > 3 else "<your-document-url>"
    model_id = "prebuilt-invoice";

    document_analysis_client = DocumentAnalysisClient(endpoint=endpoint, credential=AzureKeyCredential(key))
    poller = document_analysis_client.begin_analyze_document_from_url(model_id, doc_url)
    result = poller.result()
    for document in result.documents:
        if document.doc_type == "invoice":
            process_invoice(document)


def process_invoice(document):
    print(f"Document: {document.doc_type}")

    customer_name_field = document.fields.get("CustomerName")
    if customer_name_field:
        print(f"    CustomerName: Value={customer_name_field.value}   Confidence={customer_name_field.confidence}")
    
    customer_id_field = document.fields.get("CustomerId")
    if customer_id_field:
        print(f"    CustomerId: Value={customer_id_field.value}   Confidence={customer_id_field.confidence}")
    
    purchase_order_field = document.fields.get("PurchaseOrder")
    if purchase_order_field:
        print(f"    PurchaseOrder: Value={purchase_order_field.value}   Confidence={purchase_order_field.confidence}")
    
    invoice_id_field = document.fields.get("InvoiceId")
    if invoice_id_field:
        print(f"    InvoiceId: Value={invoice_id_field.value}   Confidence={invoice_id_field.confidence}")
    
    invoice_date_field = document.fields.get("InvoiceDate")
    if invoice_date_field:
        print(f"    InvoiceDate: Value={invoice_date_field.value}   Confidence={invoice_date_field.confidence}")
    
    due_date_field = document.fields.get("DueDate")
    if due_date_field:
        print(f"    DueDate: Value={due_date_field.value}   Confidence={due_date_field.confidence}")
    
    vendor_name_field = document.fields.get("VendorName")
    if vendor_name_field:
        print(f"    VendorName: Value={vendor_name_field.value}   Confidence={vendor_name_field.confidence}")
    
    vendor_address_field = document.fields.get("VendorAddress")
    if vendor_address_field:
        print(f"    VendorAddress: Value={vendor_address_field.value}   Confidence={vendor_address_field.confidence}")
    
    vendor_address_recipient_field = document.fields.get("VendorAddressRecipient")
    if vendor_address_recipient_field:
        print(f"    VendorAddressRecipient: Value={vendor_address_recipient_field.value}   Confidence={vendor_address_recipient_field.confidence}")
    
    customer_address_field = document.fields.get("CustomerAddress")
    if customer_address_field:
        print(f"    CustomerAddress: Value={customer_address_field.value}   Confidence={customer_address_field.confidence}")
    
    customer_address_recipient_field = document.fields.get("CustomerAddressRecipient")
    if customer_address_recipient_field:
        print(f"    CustomerAddressRecipient: Value={customer_address_recipient_field.value}   Confidence={customer_address_recipient_field.confidence}")
    
    billing_address_field = document.fields.get("BillingAddress")
    if billing_address_field:
        print(f"    BillingAddress: Value={billing_address_field.value}   Confidence={billing_address_field.confidence}")
    
    billing_address_recipient_field = document.fields.get("BillingAddressRecipient")
    if billing_address_recipient_field:
        print(f"    BillingAddressRecipient: Value={billing_address_recipient_field.value}   Confidence={billing_address_recipient_field.confidence}")
    
    shipping_address_field = document.fields.get("ShippingAddress")
    if shipping_address_field:
        print(f"    ShippingAddress: Value={shipping_address_field.value}   Confidence={shipping_address_field.confidence}")
    
    shipping_address_recipient_field = document.fields.get("ShippingAddressRecipient")
    if shipping_address_recipient_field:
        print(f"    ShippingAddressRecipient: Value={shipping_address_recipient_field.value}   Confidence={shipping_address_recipient_field.confidence}")
    
    sub_total_field = document.fields.get("SubTotal")
    if sub_total_field:
        print(f"    SubTotal: Value={sub_total_field.value}   Confidence={sub_total_field.confidence}")
    
    total_tax_field = document.fields.get("TotalTax")
    if total_tax_field:
        print(f"    TotalTax: Value={total_tax_field.value}   Confidence={total_tax_field.confidence}")
    
    invoice_total_field = document.fields.get("InvoiceTotal")
    if invoice_total_field:
        print(f"    InvoiceTotal: Value={invoice_total_field.value}   Confidence={invoice_total_field.confidence}")
    
    amount_due_field = document.fields.get("AmountDue")
    if amount_due_field:
        print(f"    AmountDue: Value={amount_due_field.value}   Confidence={amount_due_field.confidence}")
    
    previous_unpaid_balance_field = document.fields.get("PreviousUnpaidBalance")
    if previous_unpaid_balance_field:
        print(f"    PreviousUnpaidBalance: Value={previous_unpaid_balance_field.value}   Confidence={previous_unpaid_balance_field.confidence}")
    
    remittance_address_field = document.fields.get("RemittanceAddress")
    if remittance_address_field:
        print(f"    RemittanceAddress: Value={remittance_address_field.value}   Confidence={remittance_address_field.confidence}")
    
    remittance_address_recipient_field = document.fields.get("RemittanceAddressRecipient")
    if remittance_address_recipient_field:
        print(f"    RemittanceAddressRecipient: Value={remittance_address_recipient_field.value}   Confidence={remittance_address_recipient_field.confidence}")
    
    service_address_field = document.fields.get("ServiceAddress")
    if service_address_field:
        print(f"    ServiceAddress: Value={service_address_field.value}   Confidence={service_address_field.confidence}")
    
    service_address_recipient_field = document.fields.get("ServiceAddressRecipient")
    if service_address_recipient_field:
        print(f"    ServiceAddressRecipient: Value={service_address_recipient_field.value}   Confidence={service_address_recipient_field.confidence}")
    
    service_start_date_field = document.fields.get("ServiceStartDate")
    if service_start_date_field:
        print(f"    ServiceStartDate: Value={service_start_date_field.value}   Confidence={service_start_date_field.confidence}")
    
    service_end_date_field = document.fields.get("ServiceEndDate")
    if service_end_date_field:
        print(f"    ServiceEndDate: Value={service_end_date_field.value}   Confidence={service_end_date_field.confidence}")
    
    total_v_a_t_field = document.fields.get("TotalVAT")
    if total_v_a_t_field:
        print(f"    TotalVAT: Value={total_v_a_t_field.value}   Confidence={total_v_a_t_field.confidence}")
    
    vendor_tax_id_field = document.fields.get("VendorTaxId")
    if vendor_tax_id_field:
        print(f"    VendorTaxId: Value={vendor_tax_id_field.value}   Confidence={vendor_tax_id_field.confidence}")
    
    customer_tax_id_field = document.fields.get("CustomerTaxId")
    if customer_tax_id_field:
        print(f"    CustomerTaxId: Value={customer_tax_id_field.value}   Confidence={customer_tax_id_field.confidence}")
    
    payment_term_field = document.fields.get("PaymentTerm")
    if payment_term_field:
        print(f"    PaymentTerm: Value={payment_term_field.value}   Confidence={payment_term_field.confidence}")
    
    items_field = document.fields.get("Items")
    if items_field:
        for index, item in enumerate(items_field.value):
            item_fields = item.value
            
            amount_field = item_fields.get("Amount")
            if amount_field:
                print(f"    Items[{index}].Amount: Value={amount_field.value}   Confidence={amount_field.confidence}")
            
            date_field = item_fields.get("Date")
            if date_field:
                print(f"    Items[{index}].Date: Value={date_field.value}   Confidence={date_field.confidence}")
            
            description_field = item_fields.get("Description")
            if description_field:
                print(f"    Items[{index}].Description: Value={description_field.value}   Confidence={description_field.confidence}")
            
            quantity_field = item_fields.get("Quantity")
            if quantity_field:
                print(f"    Items[{index}].Quantity: Value={quantity_field.value}   Confidence={quantity_field.confidence}")
            
            product_code_field = item_fields.get("ProductCode")
            if product_code_field:
                print(f"    Items[{index}].ProductCode: Value={product_code_field.value}   Confidence={product_code_field.confidence}")
            
            tax_field = item_fields.get("Tax")
            if tax_field:
                print(f"    Items[{index}].Tax: Value={tax_field.value}   Confidence={tax_field.confidence}")
            
            unit_field = item_fields.get("Unit")
            if unit_field:
                print(f"    Items[{index}].Unit: Value={unit_field.value}   Confidence={unit_field.confidence}")
            
            unit_price_field = item_fields.get("UnitPrice")
            if unit_price_field:
                print(f"    Items[{index}].UnitPrice: Value={unit_price_field.value}   Confidence={unit_price_field.confidence}")
            
            v_a_t_field = item_fields.get("VAT")
            if v_a_t_field:
                print(f"    Items[{index}].VAT: Value={v_a_t_field.value}   Confidence={v_a_t_field.confidence}")
            
    
if __name__ == "__main__":
    analyze_document(sys.argv)

File: articles/applied-ai-services/form-recognizer/prebuilt-models/sample-code/csharp/prebuilt-invoice.2022-06-30-preview.cs
using System;
using Azure;
using Azure.AI.FormRecognizer.DocumentAnalysis;

namespace FormRecognizerSample
{
    internal class Program
    {
        static void Main(string[] args)
        {
            var endpoint = args.Length > 0 ? args[0] : "<your-endpoint>";
            var key = args.Length > 1 ? args[1] : "<your-key>";
            var documentUrl = args.Length > 2 ? args[2] : "<your-document-url>";
            var modelId = "prebuilt-invoice";

            var client = new DocumentAnalysisClient(new Uri(endpoint), new AzureKeyCredential(key));
            var operation = client.StartAnalyzeDocumentFromUri(modelId, new Uri(documentUrl));
            var result = operation.WaitForCompletion().Value;
            foreach (var document in result.Documents)
            {
                switch (document.DocType)
                {
                    case "invoice": ProcessInvoice(document); break;
                }
            }
        }

        static void ProcessInvoice(AnalyzedDocument document)
        {
            Console.WriteLine($"Document: {document.DocType}");
            if (document.Fields.TryGetValue("CustomerName", out DocumentField? customerNameField))
            {
                Console.WriteLine($"  CustomerName: Value={customerNameField.AsString()}   Confidence={customerNameField.Confidence}");
            }
            if (document.Fields.TryGetValue("CustomerId", out DocumentField? customerIdField))
            {
                Console.WriteLine($"  CustomerId: Value={customerIdField.AsString()}   Confidence={customerIdField.Confidence}");
            }
            if (document.Fields.TryGetValue("PurchaseOrder", out DocumentField? purchaseOrderField))
            {
                Console.WriteLine($"  PurchaseOrder: Value={purchaseOrderField.AsString()}   Confidence={purchaseOrderField.Confidence}");
            }
            if (document.Fields.TryGetValue("InvoiceId", out DocumentField? invoiceIdField))
            {
                Console.WriteLine($"  InvoiceId: Value={invoiceIdField.AsString()}   Confidence={invoiceIdField.Confidence}");
            }
            if (document.Fields.TryGetValue("InvoiceDate", out DocumentField? invoiceDateField))
            {
                Console.WriteLine($"  InvoiceDate: Value={invoiceDateField.AsDate()}   Confidence={invoiceDateField.Confidence}");
            }
            if (document.Fields.TryGetValue("DueDate", out DocumentField? dueDateField))
            {
                Console.WriteLine($"  DueDate: Value={dueDateField.AsDate()}   Confidence={dueDateField.Confidence}");
            }
            if (document.Fields.TryGetValue("VendorName", out DocumentField? vendorNameField))
            {
                Console.WriteLine($"  VendorName: Value={vendorNameField.AsString()}   Confidence={vendorNameField.Confidence}");
            }
            if (document.Fields.TryGetValue("VendorAddress", out DocumentField? vendorAddressField))
            {
                Console.WriteLine($"  VendorAddress: Value={vendorAddressField.AsString()}   Confidence={vendorAddressField.Confidence}");
            }
            if (document.Fields.TryGetValue("VendorAddressRecipient", out DocumentField? vendorAddressRecipientField))
            {
                Console.WriteLine($"  VendorAddressRecipient: Value={vendorAddressRecipientField.AsString()}   Confidence={vendorAddressRecipientField.Confidence}");
            }
            if (document.Fields.TryGetValue("CustomerAddress", out DocumentField? customerAddressField))
            {
                Console.WriteLine($"  CustomerAddress: Value={customerAddressField.AsString()}   Confidence={customerAddressField.Confidence}");
            }
            if (document.Fields.TryGetValue("CustomerAddressRecipient", out DocumentField? customerAddressRecipientField))
            {
                Console.WriteLine($"  CustomerAddressRecipient: Value={customerAddressRecipientField.AsString()}   Confidence={customerAddressRecipientField.Confidence}");
            }
            if (document.Fields.TryGetValue("BillingAddress", out DocumentField? billingAddressField))
            {
                Console.WriteLine($"  BillingAddress: Value={billingAddressField.AsString()}   Confidence={billingAddressField.Confidence}");
            }
            if (document.Fields.TryGetValue("BillingAddressRecipient", out DocumentField? billingAddressRecipientField))
            {
                Console.WriteLine($"  BillingAddressRecipient: Value={billingAddressRecipientField.AsString()}   Confidence={billingAddressRecipientField.Confidence}");
            }
            if (document.Fields.TryGetValue("ShippingAddress", out DocumentField? shippingAddressField))
            {
                Console.WriteLine($"  ShippingAddress: Value={shippingAddressField.AsString()}   Confidence={shippingAddressField.Confidence}");
            }
            if (document.Fields.TryGetValue("ShippingAddressRecipient", out DocumentField? shippingAddressRecipientField))
            {
                Console.WriteLine($"  ShippingAddressRecipient: Value={shippingAddressRecipientField.AsString()}   Confidence={shippingAddressRecipientField.Confidence}");
            }
            if (document.Fields.TryGetValue("SubTotal", out DocumentField? subTotalField))
            {
                Console.WriteLine($"  SubTotal: Value={subTotalField.AsCurrency()}   Confidence={subTotalField.Confidence}");
            }
            if (document.Fields.TryGetValue("TotalTax", out DocumentField? totalTaxField))
            {
                Console.WriteLine($"  TotalTax: Value={totalTaxField.AsCurrency()}   Confidence={totalTaxField.Confidence}");
            }
            if (document.Fields.TryGetValue("InvoiceTotal", out DocumentField? invoiceTotalField))
            {
                Console.WriteLine($"  InvoiceTotal: Value={invoiceTotalField.AsCurrency()}   Confidence={invoiceTotalField.Confidence}");
            }
            if (document.Fields.TryGetValue("AmountDue", out DocumentField? amountDueField))
            {
                Console.WriteLine($"  AmountDue: Value={amountDueField.AsCurrency()}   Confidence={amountDueField.Confidence}");
            }
            if (document.Fields.TryGetValue("PreviousUnpaidBalance", out DocumentField? previousUnpaidBalanceField))
            {
                Console.WriteLine($"  PreviousUnpaidBalance: Value={previousUnpaidBalanceField.AsCurrency()}   Confidence={previousUnpaidBalanceField.Confidence}");
            }
            if (document.Fields.TryGetValue("RemittanceAddress", out DocumentField? remittanceAddressField))
            {
                Console.WriteLine($"  RemittanceAddress: Value={remittanceAddressField.AsString()}   Confidence={remittanceAddressField.Confidence}");
            }
            if (document.Fields.TryGetValue("RemittanceAddressRecipient", out DocumentField? remittanceAddressRecipientField))
            {
                Console.WriteLine($"  RemittanceAddressRecipient: Value={remittanceAddressRecipientField.AsString()}   Confidence={remittanceAddressRecipientField.Confidence}");
            }
            if (document.Fields.TryGetValue("ServiceAddress", out DocumentField? serviceAddressField))
            {
                Console.WriteLine($"  ServiceAddress: Value={serviceAddressField.AsString()}   Confidence={serviceAddressField.Confidence}");
            }
            if (document.Fields.TryGetValue("ServiceAddressRecipient", out DocumentField? serviceAddressRecipientField))
            {
                Console.WriteLine($"  ServiceAddressRecipient: Value={serviceAddressRecipientField.AsString()}   Confidence={serviceAddressRecipientField.Confidence}");
            }
            if (document.Fields.TryGetValue("ServiceStartDate", out DocumentField? serviceStartDateField))
            {
                Console.WriteLine($"  ServiceStartDate: Value={serviceStartDateField.AsDate()}   Confidence={serviceStartDateField.Confidence}");
            }
            if (document.Fields.TryGetValue("ServiceEndDate", out DocumentField? serviceEndDateField))
            {
                Console.WriteLine($"  ServiceEndDate: Value={serviceEndDateField.AsDate()}   Confidence={serviceEndDateField.Confidence}");
            }
            if (document.Fields.TryGetValue("TotalVAT", out DocumentField? totalVATField))
            {
                Console.WriteLine($"  TotalVAT: Value={totalVATField.AsCurrency()}   Confidence={totalVATField.Confidence}");
            }
            if (document.Fields.TryGetValue("VendorTaxId", out DocumentField? vendorTaxIdField))
            {
                Console.WriteLine($"  VendorTaxId: Value={vendorTaxIdField.AsString()}   Confidence={vendorTaxIdField.Confidence}");
            }
            if (document.Fields.TryGetValue("CustomerTaxId", out DocumentField? customerTaxIdField))
            {
                Console.WriteLine($"  CustomerTaxId: Value={customerTaxIdField.AsString()}   Confidence={customerTaxIdField.Confidence}");
            }
            if (document.Fields.TryGetValue("PaymentTerm", out DocumentField? paymentTermField))
            {
                Console.WriteLine($"  PaymentTerm: Value={paymentTermField.AsString()}   Confidence={paymentTermField.Confidence}");
            }
            if (document.Fields.TryGetValue("Items", out DocumentField? itemsField))
            {
                var index = 0;
                foreach (var item in itemsField.AsList())
                {
                    var itemFields = item.AsDictionary();
                    if (itemFields.TryGetValue("Amount", out DocumentField? amountField))
                    {
                        Console.WriteLine($"  Items[{index}].Amount: Value={amountField.AsCurrency()}   Confidence={amountField.Confidence}");
                    }
                    if (itemFields.TryGetValue("Date", out DocumentField? dateField))
                    {
                        Console.WriteLine($"  Items[{index}].Date: Value={dateField.AsDate()}   Confidence={dateField.Confidence}");
                    }
                    if (itemFields.TryGetValue("Description", out DocumentField? descriptionField))
                    {
                        Console.WriteLine($"  Items[{index}].Description: Value={descriptionField.AsString()}   Confidence={descriptionField.Confidence}");
                    }
                    if (itemFields.TryGetValue("Quantity", out DocumentField? quantityField))
                    {
                        Console.WriteLine($"  Items[{index}].Quantity: Value={quantityField.AsDouble()}   Confidence={quantityField.Confidence}");
                    }
                    if (itemFields.TryGetValue("ProductCode", out DocumentField? productCodeField))
                    {
                        Console.WriteLine($"  Items[{index}].ProductCode: Value={productCodeField.AsString()}   Confidence={productCodeField.Confidence}");
                    }
                    if (itemFields.TryGetValue("Tax", out DocumentField? taxField))
                    {
                        Console.WriteLine($"  Items[{index}].Tax: Value={taxField.AsCurrency()}   Confidence={taxField.Confidence}");
                    }
                    if (itemFields.TryGetValue("Unit", out DocumentField? unitField))
                    {
                        Console.WriteLine($"  Items[{index}].Unit: Value={unitField.AsString()}   Confidence={unitField.Confidence}");
                    }
                    if (itemFields.TryGetValue("UnitPrice", out DocumentField? unitPriceField))
                    {
                        Console.WriteLine($"  Items[{index}].UnitPrice: Value={unitPriceField.AsCurrency()}   Confidence={unitPriceField.Confidence}");
                    }
                    if (itemFields.TryGetValue("VAT", out DocumentField? vATField))
                    {
                        Console.WriteLine($"  Items[{index}].VAT: Value={vATField.AsCurrency()}   Confidence={vATField.Confidence}");
                    }
                    index++;
                }
            }
        }
    }
}